Create composite action to create constructor-based installers

marcoesters opened the pull request:
This PR creates a composite action to build `constructor`-based installers. The goal of this action is to unify the build workflows for both Anaconda and Miniforge installers and pool maintenance resources.

The action takes or creates an `environment.yaml` to create a build environment. Inside that build environment, it runs `constructor` on a provided recipe directory to create the installer. The output directory contains all `constructor`-created build artifacts.

The workflow is intended to be flexible and can:

* Run the creation directly on a runner or inside a Docker container.
* Use GitHub expressions to dynamically set the content of the build environment.
* Can use existing `conda` installations on the runner/image or use `conda-standalone`/`micromamba` to create the build environment.
* Supports `constructor`'s interface for signing installers.

The README file contains several use cases and I added tests for each of them.

Closes #3

jaimergp commented on .github/workflows/test-defaults.yaml:
```diff
@@ -0,0 +1,98 @@
+          fi
+          . ${CONDA}/etc/profile.d/conda.sh
+          conda activate
+          conda info --json | python -c "import json, os, sys; from pathlib import Path; info = json.loads(sys.stdin.read()); assert Path(os.environ.get('INSTALL_DIR', '')) == Path(info['root_prefix'])"
```
```suggestion
          conda info --json | python -c "import json, os, sys; from pathlib import Path; info = json.loads(sys.stdin.read()); assert Path(os.environ.get('INSTALL_DIR', '')).samefile(info['root_prefix'])"
```

jaimergp commented on .github/workflows/test-micromamba.yaml:
```diff
@@ -0,0 +1,94 @@
+        run: |
+          . ${CONDA}/etc/profile.d/conda.sh
+          conda activate
+          conda info --json | python -c "import json, os, sys; from pathlib import Path; info = json.loads(sys.stdin.read()); assert Path(os.environ.get('INSTALL_DIR', '')) == Path(info['root_prefix'])"
```
```suggestion
          conda info --json | python -c "import json, os, sys; from pathlib import Path; info = json.loads(sys.stdin.read()); assert Path(os.environ.get('INSTALL_DIR', '')).samefile(info['root_prefix'])"
```

jaimergp commented on action.yaml:
```diff
@@ -0,0 +1,196 @@
+  environment-yaml-file:
+    description: |
+      Path to the environment.yaml file used to create the build environment.
+      Takes precedence over envrionemnt-yaml-string.
```
```suggestion
      Takes precedence over environment-yaml-string.
```

jaimergp commented on action.yaml:
```diff
@@ -0,0 +1,196 @@
+      build environment. Can be used to include logic and expressions from the calling workflow.
+      The file must not have the `name` property.
+    required: false
+  recipe-directory:
+    description: The directory containing the recipe file.
+    required: true
```
I wouldn't call the `construct.yaml` a "recipe". It can be confusing because this terminology is used for conda-build and so on. What about `input-directory` or `construct-directory` instead?

jaimergp commented on action.yaml:
```diff
@@ -0,0 +1,196 @@
+    description: The location of the standalone binary executable.
+    required: false
+
+  # secrets needed by constructor
+  azure-signtool-key-vault-client-id:
+    description: Client ID of the Azure key vault.
+    required: false
+  azure-signtool-key-vault-certificate:
+    description: Certificate of the Azure key vault.
+    required: false
+  azure-signtool-key-vault-secret:
+    description: Secret for the Azure key vault.
+    required: false
+  azure-signtool-key-vault-tenant-id:
+    description: Tenant ID of the Azure key vault.
+    required: false
+  azure-signtool-key-vault-url:
+    description: URL to for the Azure key vault.
+    required: false
+  constructor-pfx-certificate-password:
+    description: Windows signing password.
+    required: false
```
Is there a way to mark these as secrets? I thought there was a way but I could only find that for [reusable workflows](https://docs.github.com/en/actions/sharing-automations/reusing-workflows#using-inputs-and-secrets-in-a-reusable-workflow), not actions? 🤔

jaimergp commented on action.yaml:
```diff
@@ -0,0 +1,196 @@
+            CONSTRUCTOR_WORKSPACE=$(cygpath "${CONSTRUCTOR_WORKSPACE}")
+        fi
+        mkdir -p "${CONSTRUCTOR_WORKSPACE}"
+        chmod 777 "${CONSTRUCTOR_WORKSPACE}"
```
Do we really need 777?

marcoesters replied:
Unfortunately, yes. For builds inside a container, the UID and GIDs are not the same for the runner and the container image, and you need `x` permissions to access that directory. If we want to mitigate the risks here somewhat, we can set those permissions only for those builds that use a container image.

marcoesters replied:
I changed it so that the default permissions are restored in the end to further mitigate risk.

jaimergp replied:
That step won't run if there's a failure. Added a comment below.

jaimergp commented on action.yaml:
```diff
@@ -0,0 +1,196 @@
+runs:
+  using: composite
+  steps:
+    - name: Validate input
+      run: |
+        if [[ -z "${{ inputs.environment-yaml-file }}" ]] && [[ -z "${{ inputs.environment-yaml-string }}" ]]; then
+          echo "::error::\`environment-yaml-file\` or \`environment-yaml-string\` must be set."
+          exit 1
+        fi
+      shell: bash
```
I don't fully get why these are required. It's "just" to provide a `constructor` environment, right? I think we should have a sensible default here, but do allow users to override if needed.

jaimergp replied:
I like this.

jaimergp commented on action.yaml:
```diff
@@ -0,0 +1,196 @@
+      uses: docker/setup-qemu-action@53851d14592bedcffcf25ea515637cff71ef929a     # v3.3.0
+      with:
+        # https://github.com/docker/setup-qemu-action/issues/188#issuecomment-2604322104
+        image: tonistiigi/binfmt:qemu-v8.1.5
```
Can we use the hash here, instead of the tag? Trying to minimize supply chain risks.

jaimergp commented on action.yaml:
```diff
@@ -0,0 +1,196 @@
+        CONSTRUCTOR_PFX_CERTIFICATE_PASSWORD: ${{ inputs.constructor-pfx-certificate_password }}
+      run: |
+        echo "::group::Create environment"
+        # Define variables here instead of in an enviornment to simplify Docker command
```
```suggestion
        # Define variables here instead of in an environment to simplify Docker command
```

jaimergp commented on action.yaml:
```diff
@@ -0,0 +1,196 @@
+        if [[ -n "${STANDALONE_EXE}" ]]; then
+            EXTRA_CONSTRUCTOR_ARGS+=" --conda-exe ${STANDALONE_EXE}"
+        fi
+        ${CONDA_BIN} run -p "${PREFIX}" \
```
Pass the flag here to _not_ capture output and instead live stream stdout/stderr. Not sure if it's the same for both conda & micromamba.

jaimergp commented:
Would be interesting to run https://github.com/woodruffw/zizmor on this repository to catch some potential security risks.

marcoesters commented on action.yaml:
```diff
@@ -0,0 +1,196 @@
+    description: |
+      Path to the environment.yaml file used to create the build environment.
+      Takes precedence over envrionemnt-yaml-string.
+      The file must not have the `name` property.
```
I think I mixed something up here - `name` can be used in the file, so I removed the comment.

marcoesters commented:
> Would be interesting to run https://github.com/woodruffw/zizmor on this repository to catch some potential security risks.

Thank you for your review! I think I addressed your specific comments. Also thanks for recommending zizmor - I ran the checks and made several recommended changes. The chief concern was template injection vulnerabilities, so I replaced most with environment variables.

The reason why I employed template substitutions is that adding environment variables into the `env` context of the build step is not enough - they also need to be added to building the Docker command. If we ever support other containerization software like podman, they need to be handled in that way, too. This can be quite brittle to maintain because this is easy to miss.

To mitigate template injection vulnerabilities, I added an input validation step and added tests that try and inject a `cat` command output (see [this run](https://github.com/marcoesters/installer/actions/runs/14891810348/job/41825406871) as an example for how it would succeed if we just did template substitution). The workflow will error out if those inputs do not point to a directory/file, so an attacked will have to find a file with the exact same path as the malicious code that is being executed.

To be specific, these are the templates in the build step `zizmor` still annotates (with severity):

* `steps.workspace.outputs.constructor-workspace` (informational): This is based on environment variables and code inside the action that is not controlled by the caller.
* `steps.workspace.outputs.outputs-file` (informational): Same as `constructor-workspace`.
* `inputs.environment-yaml-file` (high): mitigated by input validation
* `steps.environment-str.outputs.environment-yaml-file` (informational): same as `constructor-workspace`
* `inputs.conda-root` (high): mitigated by input validation
* `inputs.input-directory` (high): mitigated by input validation
* `inputs.standalone-location` (high): mitigated by input validation

If you see risks despite the input validation, we will have to accept the more cumbersome maintenance model though. The input validation is a good addition either way.

jaimergp commented on action.yaml:
```diff
@@ -233,3 +245,16 @@ runs:
       shell: bash
+
+    - name: Fix permissions and ownership for files created with a container
+      if: ${{ inputs.container-image }}
```
We might need some sort of `and always()` or something like that to ensure this always runs. Alternatively, you could use something like https://github.com/gacts/run-and-post-run to make this step a `post` task of the above step.

jaimergp commented on action.yaml:
```diff
@@ -0,0 +1,262 @@
+            echo "::error::File ${STANDALONE_LOCATION} does not exist."
+            exit 1
+        fi
+      shell: bash
```
If it can't be global because it's composite, then we need to change to `bash -e` in all steps.

```suggestion
      shell: bash -e
```
